### PR TITLE
[MIG][10.0] hr_timesheet_holidays

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
Standard migration

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
Please, check travis @aheficent

File: hr_timesheet_holidays/README.rst
Comment (chienandalu) [inline hr_timesheet_holidays/README.rst]: Could you develop a more detailed explanation?

File: hr_timesheet_holidays/__manifest__.py
Comment (chienandalu) [inline hr_timesheet_holidays/__manifest__.py]: OCA guidelines recommend now to set `website` to the OCA repo url: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#modules

File: hr_timesheet_holidays/__manifest__.py
Comment (chienandalu) [inline hr_timesheet_holidays/__manifest__.py]: Better one line per contributor

File: hr_timesheet_holidays/models/__init__.py
Comment (chienandalu) [inline hr_timesheet_holidays/models/__init__.py]: There's no real need for authorship in inits

File: hr_timesheet_holidays/wizard/holidays_import_view.xml
Comment (adrienpeiffer) [inline hr_timesheet_holidays/wizard/holidays_import_view.xml]: You should put those buttons in the footer. See attach screenshot.
timesheet_holidays01

**Review by adrienpeiffer — CHANGES_REQUESTED** [review]

File: hr_timesheet_holidays/wizard/holidays_import_view.xml
Comment (AaronHForgeFlow) [reply hr_timesheet_holidays/wizard/holidays_import_view.xml]: You're right @adrienpeiffer Thanks

File: hr_timesheet_holidays/README.rst
Comment (elicoidal) [inline hr_timesheet_holidays/README.rst]: one line before

File: hr_timesheet_holidays/README.rst
Comment (elicoidal) [inline hr_timesheet_holidays/README.rst]: svg ->png

File: hr_timesheet_holidays/wizard/holidays_import.py
Comment (elicoidal) [inline hr_timesheet_holidays/wizard/holidays_import.py]: I am not fan of SQL statements in the code: could we use ORM functions instead?

File: hr_timesheet_holidays/wizard/holidays_import.py
Comment (elicoidal) [inline hr_timesheet_holidays/wizard/holidays_import.py]: I understand the idea of a wizard but why not create the entry in the analytic account when the holiday is validated?
Would there be any reason it would not be possible?

**Review by elicoidal — CHANGES_REQUESTED** [review]
small details. waiting for runbot to test

**Comment by elicoidal** [comment]
I am not sure I understand well the feature and how it has been implemented. I would have expected the following use case:
1. set up of the Leaves with AA
2. User creates and validate leave request (eg 1.5 day)
3. User launches the wizard in timesheet report
4. a new AA line is added to the current AA with 8h on the first day, 4h in second)
5. I can check in the first tab of the timesheet report the time work in project and the time on leave.

Currently: No line is added and the result in the report is not straightforward (check current runbot on Pieter Parker).

I come back to my previous remark: I would rather create the line in the Timesheet report directly when the leave is validated (it could be with a checkbox at leave level if wizard behavior is still needed).

**Comment by feketemihai** [comment]
@aheficent The module is merged on v8, so probably we should start from it, since it's working fine, no wizard, sql statements...

**Comment by AaronHForgeFlow** [comment]
@elicoidal The wizard behavior is not really needed so it is removed now. The timesheet is updated as soon as the leave is approved. Setting up the AA as leave account is needed.  Thanks for reviewing.

**Comment by elicoidal** [comment]
@aheficent it looks like something is not yet there (or my use case is not correct)
1. I created new leave type and AA:
screenshot from 2018-04-05 11-01-55
2. I created 2 leaves (one on not created timesheet report and one on the current week timesheet report)
screenshot from 2018-04-05 11-00-09
3. I cannot see neither of AA lines in the timesheets reports after validating the leaves:
screenshot from 2018-04-05 11-03-17

Did I forget something?

**Comment by feketemihai** [comment]
@elicoidal The issue is that in v10, the allow_timesheets fields is only set to projects, not to analytic.account, just like in v8, there in hr_timesheet module the fiels was named use_timesheets and was set to analytic.account...so we should adapt this @aheficent

**Comment by AaronHForgeFlow** [comment]
@feketemihai you're right. I'll do the thing

**Comment by AaronHForgeFlow** [comment]
A new constraint has been set. This is avoiding to set accounts whose projects does not allow timesheets. However I'm not sure if this is correct. Maybe it's better to allow to set leave accounts where only one of the projects related to the account allow_timesheets. What do you think?

**Comment by feketemihai** [comment]
@aheficent I don't think the constrain have something to do with the non working module, we should not use the constrain because these analytic accounts were not meant to be linked with a project. even with the constrain or not the only issue that we have is that the analytic lines are created but they are not linked to a sheet, because of this https://github.com/odoo/odoo/blob/10.0/addons/hr_timesheet_sheet/models/account_analytic_line.py#L15, which we should modify to allow if the line have a analytic account and that analytic account is a leave account.

File: hr_timesheet_holiday/models/hr_holidays.py
Comment (feketemihai) [inline hr_timesheet_holiday/models/hr_holidays.py]: @aheficent Probably here to add a condition about states of the sheet, just like in _compute_sheet from account.analytic.line, https://github.com/odoo/odoo/blob/10.0/addons/hr_timesheet_sheet/models/account_analytic_line.py#L15

File: hr_timesheet_holiday/models/hr_holidays.py
Comment (feketemihai) [inline hr_timesheet_holiday/models/hr_holidays.py]: remove this since the sheet_id is computed on analytic lines

**Review by feketemihai — CHANGES_REQUESTED** [review]

**Comment by AaronHForgeFlow** [comment]
@feketemihai I tried to avoid havinfg analytic lines not attached to any timesheet here:
https://github.com/Eficent/hr-timesheet/blob/10.0-mig-hr_timesheet_holidays/hr_timesheet_holiday/models/hr_holidays.py#L90

It's not doing the thing then

**Comment by feketemihai** [comment]
@aheficent The error is raised if you don't have a sheet for that period, so you cannot validate the leave, but if you have a active sheet, the analytic lines are created but they are not linked to a sheet, because of sheet_id field in account.analytic.line which is a computed one, but it's computed only when the line is linked to a project. So, we will have to allow the computation of sheet_id if we have an analytic account set on the line, and that analytic account is a leave one.

**Comment by feketemihai** [comment]
@api.depends('date', 'user_id', 'project_id', 'sheet_id_computed.date_to', 'sheet_id_computed.date_from', 'sheet_id_computed.employee_id')
    def _compute_sheet(self):
        """Links the timesheet line to the corresponding sheet
        """
        super(AnalyticLine, self)._compute_sheet()
        for ts_line in self:
            if ts_line.leave_id:
                continue
            sheets = self.env['hr_timesheet_sheet.sheet'].search(
                [('date_to', '>=', ts_line.date), ('date_from', '<=', ts_line.date),
                 ('employee_id.user_id.id', '=', ts_line.user_id.id),
                 ('state', 'in', ['draft', 'new'])])
            if sheets:
                # [0] because only one sheet possible for an employee between 2 dates
                ts_line.sheet_id_computed = sheets[0]
                ts_line.sheet_id = sheets[0]

**Comment by feketemihai** [comment]
@aheficent Thanks for all your qork, but i still have one minor request, can you check the view on timesheets, because in the view is only the project_id field, so every leave will be showed with False, probably to add the analytic account after the project, to have both the view.

**Comment by AaronHForgeFlow** [comment]
@feketemihai Sure, I will check, thank you for reviewing!

**Comment by AaronHForgeFlow** [comment]
It will still show `False` in the `weekly_timesheet` widget. I'm trying to override it.

**Comment by AaronHForgeFlow** [comment]
@feketemihai I think that the constraint that forces to have a project related to the analytic account makes sense. The accounts are meant to be linked to a project. Indeed, in OpenUpgrade if there was any orphan analytic account, a new project is created and linked to the orphan analytic account: https://github.com/OCA/OpenUpgrade/blob/10.0/addons/hr_timesheet/migrations/10.0.1.0/post-migration.py#L21

**Comment by feketemihai** [comment]
I don't have an issue about the False in project, just to add the analytic account field in view, something like

timesheet

**Comment by AaronHForgeFlow** [comment]
I did not mean to workaround the issue, but to keep consistency in the module. If all the analytic accounts in v10 used for timesheets has a project associated, it makes sense to me to allow only analytic_accounts associated to a project for leaves accounts.

**Comment by feketemihai** [comment]
@aheficent Then i think it's better to add this in readme -> configuration, to create a new project for each analytic account that is set as leave,  but then we should change the code to complete the project_id fields as well, and no modification in the view.

**Comment by AaronHForgeFlow** [comment]
I agree, I will go for that.

**Comment by AaronHForgeFlow** [comment]
Last changes:
- add constraint to use only accounts linked to projects
- create a project linked to the account when the account is `is_leave_account`
- adapt test

Known issues:
- The analytic lines are correctly created but the widget `weekly_timesheet` is still not finding the project


Update: It works now. Ready for review.

**Comment by CasVissers-360ERP** [comment]
I got the error:
"This line is protected against editing because it was created automatically by a leave request. Please edit the leave request instead."
While no analytic line was created. I solved this by changing:
self.sudo(user.id).**with_context(force_write=True)**.write({'analytic_line_ids':

**Comment by CasVissers-360ERP** [comment]
I would also like if the _find_sheet functionallity is removed or it's made optional. For me it makes no sense that a manager should remind the user to create a timesheet or create the timesheet for the user. Additionally once the timesheet line is created it will be automatically added to new and existing timesheets, so for me it makes no sense to check on it.

**Comment by CasVissers-360ERP** [comment]
When the project related to the analytic account is archived:
```
Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/customer/odoo/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/customer/odoo/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/customer/odoo/odoo/http.py", line 333, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/customer/odoo/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/customer/odoo/odoo/http.py", line 326, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/customer/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/customer/odoo/odoo/http.py", line 506, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/customer/odoo/addons/web/controllers/main.py", line 894, in call_button
    action = self._call_kw(model, method, args, {})
  File "/opt/odoo/customer/odoo/addons/web/controllers/main.py", line 882, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/customer/odoo/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/customer/odoo/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/customer/community/hr_timesheet_holiday/models/hr_holidays.py", line 101, in action_approve
    user=user,
  File "/opt/odoo/customer/community/hr_timesheet_holiday/models/hr_holidays.py", line 33, in add_timesheet_line
    'project_id': account.project_ids[0].id,
  File "/opt/odoo/customer/odoo/odoo/models.py", line 5228, in __getitem__
    return self._browse((self._ids[key],), self.env)
IndexError: tuple index out of range
```

**Comment by AaronHForgeFlow** [comment]
I'll fix that

**Comment by rafaelbn** [comment]
@aheficent once all comments attended and fixed I recommend you to ping reviewers. Only with commits sometimes we miss them

**Review by CasVissers-360ERP — APPROVED** [review]
Approved, thank you! Sorry for the delay was on a holiday.

**Comment by AaronHForgeFlow** [comment]
@adrienpeiffer @elicoidal @feketemihai Can you please update your review? Thank you!

**Review by elicoidal — APPROVED** [review]
LGTM

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by AaronHForgeFlow** [comment]
squashed. Readme updated to show only projects are allowed. @feketemihai can you remove the WIP label? Thanks

**Comment by MiquelRForgeFlow** [comment]
@aheficent add `hr_holidays_settings` as dependency 😉

**Review by JordiBForgeFlow — CHANGES_REQUESTED** [review]
Please change the module so that leave types point to projects instead of analytic accounts. Please add a migration script similar to https://github.com/OCA/OpenUpgrade/blob/10.0/addons/hr_timesheet/migrations/10.0.1.0/post-migration.py

**Comment by MiquelRForgeFlow** [comment]
@aheficent @jbeficent leave types point to projects now, let's update review again! 😉

File: hr_timesheet_holiday/models/project.py
Comment (AaronHForgeFlow) [inline hr_timesheet_holiday/models/project.py]: I  think this is not necessary anymore

File: hr_timesheet_holiday/README.rst
Comment (AaronHForgeFlow) [inline hr_timesheet_holiday/README.rst]: select the project

**Review by AaronHForgeFlow — COMMENTED** [review]
On runbot, when confirming a leave it is not passed to the timesheet, can you check?

**Comment by MiquelRForgeFlow** [comment]
Checked. It's ready.

**Review by AaronHForgeFlow — COMMENTED** [review]
:+1: functionally tested after @mreficent changes

**Review by JordiBForgeFlow — APPROVED** [review]

**Pull request merged**
